cider/exec/operator/join/CiderLinearProbingHashtable.h

```diff
@@ -0,0 +1,273 @@
+/*
+ * Copyright (c) 2022 Intel Corporation.
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*
+A hash map for join. Uses open addressing with linear probing.
+Advantages:
+  - Predictable performance. Doesn't use the allocator unless load factor
+    grows beyond 50%. Linear probing ensures cash efficency.
+  - Desgin for no delete/erase action, makes it faster on insert and lookup
+  - Allow duplicate keys
+Disadvantages:
+  - Significant performance degradation at high load factors.
+  - Maximum load factor hard coded to 50%, memory inefficient.
+ */
+#pragma once
+
+#include <algorithm>
+#include <cstddef>
+#include <cstdint>
+#include <limits>
+#include <stdexcept>
+#include <vector>
+
+namespace CiderLinearProbing {
```

Comment:
And why put this hash table under join dir?

Reply:
The folder and dir can be discussed. Put this in join is just because it will be used for join recently .